message/messages.go

```diff
@@ -186,69 +211,123 @@ func (mb *msgBuilder) marshal(
 	//
 	// This recursive packing allows us to avoid an extra compression on/off
 	// field in the message.
-	startTime := time.Now()
-	compressedBytes, err := mb.gzipCompressor.Compress(uncompressedMsgBytes)
-	if err != nil {
-		return nil, 0, 0, err
+	var (
+		startTime     = time.Now()
+		compressedMsg p2p.Message
+	)
+	switch compressionType {
+	case compression.TypeNone:
+		return uncompressedMsgBytes, 0, op, nil
+	case compression.TypeGzip:
+		compressedBytes, err := mb.gzipCompressor.Compress(uncompressedMsgBytes)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+		compressedMsg = p2p.Message{
+			Message: &p2p.Message_CompressedGzip{
+				CompressedGzip: compressedBytes,
+			},
+		}
+
+	case compression.TypeZstd:
+		compressedBytes, err := mb.zstdCompressor.Compress(uncompressedMsgBytes)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+		compressedMsg = p2p.Message{
+			Message: &p2p.Message_CompressedZstd{
+				CompressedZstd: compressedBytes,
+			},
+		}
+	default:
+		return nil, 0, 0, errUnknownCompressionType
 	}
 
-	compressedMsg := p2p.Message{
-		Message: &p2p.Message_CompressedGzip{
-			CompressedGzip: compressedBytes,
-		},
-	}
 	compressedMsgBytes, err := proto.Marshal(&compressedMsg)
 	if err != nil {
 		return nil, 0, 0, err
 	}
 	compressTook := time.Since(startTime)
 
+	switch compressionType {
+	case compression.TypeGzip:
+		mb.gzipCompressTimeMetrics[op].Observe(float64(compressTook))
+	case compression.TypeZstd:
+		mb.zstdCompressTimeMetrics[op].Observe(float64(compressTook))
+	}
+
 	bytesSaved := len(uncompressedMsgBytes) - len(compressedMsgBytes)
-	return compressedMsgBytes, bytesSaved, compressTook, nil
+	return compressedMsgBytes, bytesSaved, op, nil
 }
 
-func (mb *msgBuilder) unmarshal(b []byte) (*p2p.Message, bool, int, time.Duration, error) {
+func (mb *msgBuilder) unmarshal(b []byte) (*p2p.Message, int, Op, error) {
 	m := new(p2p.Message)
 	if err := proto.Unmarshal(b, m); err != nil {
-		return nil, false, 0, 0, err
+		return nil, 0, 0, err
 	}
 
-	compressed := m.GetCompressedGzip()
-	if len(compressed) == 0 {
+	// Figure out what compression type, if any, was used to compress the message.
+	var (
+		compressionType compression.Type
```

Comment:
I think this is only used to switch on the metrics later - can we just switch this to the metrics map instead and avoid the lower switch?